### PR TITLE
Use music-metadata to extract audio metadata

### PR DESCRIPTION
**Pull request by Borewit**
Please give [music-metadata](https://github.com/Borewit/music-metadata) a try, it is designed for this:
* support for promises
* TypeScript definitions build in
* Works with any audio file and metadata header
* Most advanced and [frequently used](https://npmcharts.com/compare/music-metadata,jsmediatags,musicmetadata,node-id3,music-metadata-browser,id3-parser,mp4-parser,parse-audio-metadata,music-tags,id3,read-id3-tags,stream-id3,wav-file-info,audio-metadata?start=1000&interval=30) JavaScript metadata parser.

I have not tested the changes.

Good luck with your project, let me know if you [experience any issue](https://github.com/Borewit/music-metadata/issues) or face great success ;-)

### CODE REVIEW

**Comment by tvararu** [comment]
Thanks for the PR! Will review this locally later. I don’t have automated tests for this because it’s a prototype.

File: db/populate.js [hunk 1]
```diff
@@ -30,17 +32,22 @@ songs.forEach(filePath => {
         title,
         track: parseTrack(track)
       });
-      if (parsedSongs.length == songs.length) {
+    } catch(err) {
+      console.log(`Error parsing ${filePath}: ${error.message}`);
```
Comment (tvararu) [inline db/populate.js]:
```suggestion
      console.log(`Error parsing ${filePath}: ${err.message}`);
```

File: db/populate.js [hunk 2]
```diff
@@ -30,17 +32,22 @@ songs.forEach(filePath => {
...
+      if (parsedSongs.length === songs.length) {
         parsedSongs.sort(sortByArtistAlbumTrack);
         outputLibrary();
         outputGraphLibrary();
       }
-    },
-    onError: error => {
-      console.log(":(", error.type, error.info);
+    } catch(err) {
+      console.log(`Error processing metadata: ${error.message}`);
```
Comment (tvararu) [inline db/populate.js]:
```suggestion
      console.log(`Error processing metadata: ${err.message}`);
```

**Review by tvararu — COMMENTED** [review]
Unfortunately, I seem to get some errors when using this library as opposed to `jsmediatags`:

```bash
Error parsing /Users/tvararu/Music/Music/Media/Music/Tool/Fear Inoculum/1-01 Fear Inoculum.m4a: FourCC contains invalid characters: 48 b5 9b 70
Error parsing /Users/tvararu/Music/Music/Media/Music/Tool/Fear Inoculum/1-02 Pneuma.m4a: FourCC contains invalid characters: f5 12 80 b2
Error parsing /Users/tvararu/Music/Music/Media/Music/Tool/Fear Inoculum/1-04 Invincible.m4a: FourCC contains invalid characters: 25 6e 4c 3a
Error parsing /Users/tvararu/Music/Music/Media/Music/Tool/Fear Inoculum/1-08 Chocolate Chip Trip.m4a: FourCC contains invalid characters: 9d 51 82 0e
Error parsing /Users/tvararu/Music/Music/Media/Music/Tool/Fear Inoculum/1-10 Mockingbeat.m4a: FourCC contains invalid characters: 11 3d a9 c9
Error parsing /Users/tvararu/Music/Music/Media/Music/Tool/Fear Inoculum/1-06 Descending.m4a: FourCC contains invalid characters: 19 d7 00 ec
Error parsing /Users/tvararu/Music/Music/Media/Music/Tool/Fear Inoculum/1-05 Legion Inoculant.m4a: FourCC contains invalid characters: a9 ea 10 3d
Error parsing /Users/tvararu/Music/Music/Media/Music/Tool/Fear Inoculum/1-03 Litanie contre la Peur.m4a: FourCC contains invalid characters: f7 8b ef bd
Error parsing /Users/tvararu/Music/Music/Media/Music/Tool/Fear Inoculum/1-07 Culling Voices.m4a: FourCC contains invalid characters: ed ff f1 56
```

@Borewit any ideas?

**Comment by Borewit** [comment]
> Unfortunately, I seem to get some errors when using this library as opposed to `jsmediatags`:
> @Borewit any ideas?

I am sorry and surprised to hear you run into errors.
That looks very weird.
What I can see from log is that MP4 parser gets stuck on an invalid atom (box) header.

You can activate console logging if you set `DEBUG=*`;
You can also drag and drop it here: https://audio-tag-analyzer.netlify.com, and check what is says (is using the music-metadata parser).

I would like to examine one of those files.
Can you please zip one file @tvararu it and paste here, or pass it via https://wetransfer.com/?

**Comment by tvararu** [comment]
@Borewit sure, I get a similar error message on https://audio-tag-analyzer.netlify.com:

Screenshot 2019-11-10 at 11 27 26

Here is the file:

[not-working-audio-file.zip](https://github.com/tvararu/music.vararu.org/files/3828691/not-working-audio-file.zip)

Let me know if I can help in any other way. :)

**Comment by Borewit** [comment]
I am quite convinced the error is caused due to some data at the end of those files, which does not belong there. 
Or worse, the `mdat` atom has an invalid header length (too short), and part of the audio data is therefor not included.

The noise starts at offset 20517571, which is 519878‬ bytes away from the end of the file.

Why does jsmediatags not complain? It already stopped reading at that point. music-metadata scans the file much deeper, one of the reasons is that files may have appending file tags like APEv2.

Anyway I was able to play the file, so other software does not make a big fuzz on the extra data, so I decided music-metadata should not be so picky neither.

I created a PR: Borewit/music-metadata#319
I think I will add another unit test before I merge and release.

I'll keep you posted.

**Comment by tvararu** [comment]
Very good detective skills @Borewit! 🕵

We can switch to music-metadata when the updated version has released.

As a side-note, is extracting album art simple? I noticed there is a `skipCovers` option in the `README` so I assume there must be!

**Comment by Borewit** [comment]
The [issue with appending noise in MP4 files has been solved](https://github.com/Borewit/music-metadata/issues/318).
I updated to [music-metadata to v4.9.0](https://github.com/Borewit/music-metadata/releases/tag/v4.9.0).

Can please give it another try @tvararu?

> As a side-note, is extracting album art simple? I noticed there is a skipCovers option in the README so I assume there must be!

Via `metadata.common.picture` you can access an array of cover art if present.
Each picture has this interface:

```ts
/**
 * Attached picture, typically used for cover art
 */
export interface IPicture {
  /**
   * Image mime type
   */
  format: string;
  /**
   * Image data
   */
  data: Buffer;
  /**
   * Optional description
   */
  description?: string;
  /**
   * Picture type
   */
  type?: string;
}
```

To assign `img` HTML-object you can do something like:
```js
img.src = `data:${picture.format};base64,${picture.data.toString('base64')}`;
```

**Comment by tvararu** [comment]
@Borewit thanks a lot for the update! I’ll check tomorrow when I’m at my laptop. :)

--
Theodor Vararu
https://vararu.org

**Review by tvararu — APPROVED** [review]
Works well! I need to make two changes to the parsing (genre is only temporary):

```diff
@@ -28,9 +28,9 @@ const parseFilePath = filePath =>
         artist,
         album,
         year,
-        genre,
+        genre: genre[0],
         title,
-        track: parseTrack(track)
+        track: track.no
       });
     } catch(err) {
       console.log(`Error parsing ${filePath}: ${err.message}`)
```

I can handle this myself. :)

**Comment by tvararu** [comment]
Thanks a lot for your contribution @Borewit and also for the info re: album art! 🙌 

I've never had someone from open-source improve one of my toy projects out of the blue like this, you're awesome!

**Comment by Borewit** [comment]
No problem, thanks for pushing the boundaries of my toy project @tvararu.

**Pull request merged**